KeyError: 'position'

### Self Checks

- [X] This is only for bug report, if you would like to ask a question, please head to [Discussions](https://github.com/langgenius/dify/discussions/categories/general).
- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Dify version

0.9.1

### Cloud or Self Hosted

Self Hosted (Docker)

### Steps to reproduce

edge error：
![error1](https://github.com/user-attachments/assets/3b22ee0d-aa1e-4c7a-bf27-a8cf4ca4ef9e)

```
[2024-10-08 09:44:55 +0000] [41] [ERROR] Error handling request
Traceback (most recent call last):
  File "/app/api/.venv/lib/python3.10/site-packages/gunicorn/workers/base_async.py", line 115, in handle_request
    for item in respiter:
  File "/app/api/.venv/lib/python3.10/site-packages/werkzeug/wsgi.py", line 256, in __next__
    return self._next()
  File "/app/api/.venv/lib/python3.10/site-packages/werkzeug/wrappers/response.py", line 32, in _iter_encoded
    for item in iterable:
  File "/app/api/.venv/lib/python3.10/site-packages/flask/helpers.py", line 113, in generator
    yield from gen
  File "/app/api/libs/helper.py", line 185, in generate
    yield from response
  File "/app/api/core/app/features/rate_limiting/rate_limit.py", line 115, in __next__
    return next(self.generator)
  File "/app/api/core/app/apps/base_app_generate_response_converter.py", line 38, in _generate_simple_response
    for chunk in cls.convert_stream_simple_response(response):
  File "/app/api/core/app/apps/chat/generate_response_converter.py", line 111, in convert_stream_simple_response
    sub_stream_response_dict["metadata"] = cls._get_simple_metadata(metadata)
  File "/app/api/core/app/apps/base_app_generate_response_converter.py", line 84, in _get_simple_metadata
    "position": resource["position"],
KeyError: 'position'
```
### ✔️ Expected Behavior

_No response_

### ❌ Actual Behavior

_No response_

--- a/api/core/rag/entities/context_entities.py
+++ b/api/core/rag/entities/context_entities.py
@@ -1,6 +1,7 @@
 from pydantic import BaseModel
 from typing import Optional
 
+
 class DocumentContext(BaseModel):
     """
     Model class for document context.

--- a/api/core/rag/entities/context_entities.py
+++ b/api/core/rag/entities/context_entities.py
@@ -1,5 +1,6 @@
-from pydantic import BaseModel
 from typing import Optional
+
+from pydantic import BaseModel
 
 
 class DocumentContext(BaseModel):

--- a/api/core/rag/entities/context_entities.py
+++ b/api/core/rag/entities/context_entities.py
@@ -1,5 +1,5 @@
 from pydantic import BaseModel
-
+from typing import Optional
 
 class DocumentContext(BaseModel):
     """
@@ -7,4 +7,4 @@ class DocumentContext(BaseModel):
     """
 
     content: str
-    score: float
+    score: Optional[float] = None

--- a/api/core/rag/retrieval/dataset_retrieval.py
+++ b/api/core/rag/retrieval/dataset_retrieval.py
@@ -231,9 +231,7 @@ class DatasetRetrieval:
                                 source["content"] = segment.content
                             retrieval_resource_list.append(source)
         if hit_callback and retrieval_resource_list:
-            retrieval_resource_list = sorted(
-                retrieval_resource_list, key=lambda x: x.get("score"), reverse=True
-            )
+            retrieval_resource_list = sorted(retrieval_resource_list, key=lambda x: x.get("score"), reverse=True)
             for position, item in enumerate(retrieval_resource_list, start=1):
                 item["position"] = position
             hit_callback.return_retriever_resource_info(retrieval_resource_list)

--- a/api/core/rag/retrieval/dataset_retrieval.py
+++ b/api/core/rag/retrieval/dataset_retrieval.py
@@ -231,6 +231,13 @@ class DatasetRetrieval:
                                 source["content"] = segment.content
                             retrieval_resource_list.append(source)
         if hit_callback and retrieval_resource_list:
+            retrieval_resource_list = sorted(
+                retrieval_resource_list, key=lambda x: x.get("score"), reverse=True
+            )
+            position = 1
+            for item in retrieval_resource_list:
+                item["position"] = position
+                position += 1
             hit_callback.return_retriever_resource_info(retrieval_resource_list)
         if document_context_list:
             document_context_list = sorted(document_context_list, key=lambda x: x.score, reverse=True)

--- a/api/core/rag/retrieval/dataset_retrieval.py
+++ b/api/core/rag/retrieval/dataset_retrieval.py
@@ -234,10 +234,8 @@ class DatasetRetrieval:
             retrieval_resource_list = sorted(
                 retrieval_resource_list, key=lambda x: x.get("score"), reverse=True
             )
-            position = 1
-            for item in retrieval_resource_list:
+            for position, item in enumerate(retrieval_resource_list, start=1):
                 item["position"] = position
-                position += 1
             hit_callback.return_retriever_resource_info(retrieval_resource_list)
         if document_context_list:
             document_context_list = sorted(document_context_list, key=lambda x: x.score, reverse=True)

